### ISSUE TITLE
Rest API Format Convention

### ISSUE THREAD

**Issue by gmourier** [meilisearch/product#44]
# Description

An API must be usable by machines and users. It represents the interface to solve needs easily.

In order to reduce more and more the breaking changes at each release, we would like to agree on an API guideline before releasing a 1.0 version.

This concerns the naming of resources, parameters, the format of these parameters, naming conventions for enumerations such as the different status, response data blocks etc..

We want to make it easier to understand and use the APIs through simple conventions that are common to all our APIs.

The API will now be specified by OpenAPI (OAS) to allow teams to quickly understand the changes made by a new specification.

# Motivation

A great API must be convenient to use and solve users needs.

The development of a new endpoint, its improvement by contributors must be done on a solid base to facilitate their work and focus on the intrinsic value on the goal to achieve.

For developers, it is also much easier and faster to learn an API when endpoints share a common structure through conventions.

It is also important to consider this agreement as a gateway between the different teams.

Sharing the same conventional rules will allow us to communicate more efficiently between the different teams that would develop public and private APIs for our different upcoming products.

# Note on lifecycle

This format may still evolve over time but should already guide us in the right direction for much of the upcoming API development.

# Symbols terminology

🟢 Confident about this statement.
🟠 Need a working group.

**Comment by gmourier** [comment]
# Naming conventions and consistency

## Brief presentation of most used naming convention.

A 2010 [study](https://ieeexplore.ieee.org/document/5521745) confirms that developers find it easier to read the snake_case than the camelCase. Nevertheless, the increasing use of javascript seems to have introduced and accustomed developers to the use of camelCase.

`camelCase`: This method has been democratized by the Java language and is now widely used by the javascript eco-system. It consists in marking the beginning of each word by putting the first letter in a lower case. For example: camelCase, currentUser, addAttributeToGroup, etc.

`snake_case`: This method has been used for a long time in C and more recently in Ruby. The words are separated by underscore `_` allowing a compiler or interpreter to understand it as a single symbol, but allowing the human reader to separate the words in an almost natural way.

`spinal-case`: A variation of snake_case, spinal case uses short dashes `-` to separate words. The advantages and disadvantages are largely identical to snake case, except that many programming languages cannot accept it as a symbol (variable name or function).

## Endpoint naming convention

🟢 **Use the `spinal-case` convention.**

## Query parameter - Request body attribute naming convention

🟢 **Use the `camelCase` convention.**

## Response body attribute convention.

🟢 **Use the `camelCase` convention.**

**Comment by gmourier** [comment]
# Common features: Filtering on GET List of resource

By common features, we mean features that can be shared and used the same way on many endpoints. `Filtering`, `Sorting` and, `Pagination` capabilities are useful to manipulate lists to refine API results.

`Field selection` allows to explicitly request the data you need as an API user.

## Filtering

### Syntax 1 `&objectAttribute=value`

#### Pros
- Simple to understand.
- The different query parameters dedicated to filtering are documented in the Open Api Specification for the route and in the documentation.
- The errors are clearer because each query parameter can accept a set of defined variables.

#### Cons
- Dedicated query parameter to filter on an object attribute value. Does not share a convention with the `/search` route which has a true versatile filter syntax with the `filter` parameter.
- Limited to `=` operator.
- Limited to AND.
- Limited for timestamp window selection.

E.g I want to filter the list of updates `/updates` by processed status.

`&status=processed`

E.g. I want to filter the list of updates `/updates` by DocumentAdditions type.

`&type=DocumentAdditions`

E.g. I want to combine those filters.

`&status=processed&type=DocumentAdditions`

### Syntax 2 `&filter=a filter query`

#### Pros

- Use advantages from our `filter` syntax from the `search` endpoint.
- Can be really powerful with the combination of OR/AND and parenthesis usage.
- Filtering syntax is the same on all API lists.

#### Cons

- Simplicity is limited.
- Difficulty to know for the user what object fields are filterable except from API route description or API object description.

E.g I want to filter the list of updates /updates by processed status.

&filter=status=processed

E.g. I want to filter the list of updates /updates by DocumentAdditions type.

&filter=type=DocumentAdditions

E.g. I want to combine those filters.

&filter=status=processed AND type=DocumentAdditions

## Next Steps

Working group with @meilisearch/core-team to check if we can easily use the `filter` search syntax.

**Comment by gmourier** [comment]
# Common features: Sorting on GET List of resource

## Sorting

On some routes we would like to offer sorting according to the attributes we consider useful.

We therefore propose to use a `sort` attribute that will be easily plugged in as we go along on all the routes. As we will also use the `sort` field on the `/search` endpoint. We keep the same name to indicate a similar intent and thus facilitate the use of the API.

Here is the syntax we propose for sorting on numeric and alphanumeric attributes.

- Sort criterion can be separated by the `,` character.
- Sort can offer `asc` and `desc` orders.
- Multiple attribute can be sorted `attribute_1:asc,attribute_2:desc`

E.g. I want to sort the list of updates `/updates` by `duration` in descending order.

`sort=duration:desc`

**Comment by gmourier** [comment]
# Common features

##  Pagination

TBD

**Comment by gmourier** [comment]
# Common features

By common features, we mean features that can be shared and used the same way on many endpoints. Filtering, Sorting and, Pagination capabilities are useful to manipulate lists to refine API results.

Field selection allows to explicitly request the data you need as an API user.

## Field Selection

On some routes, we would like to offer the possibility to display specific `fields`.

We propose to use a `fields` parameter that will be easily plugged in as we go along on all the routes.

Here is the syntax we propose for fields selection.

- Object fields are separated by `,` character.
- It is possible to eliminate an object field or sub object from the response with a `-` operator.
- Could handle selection/removal of nested objects with the `.` character. e.g. `type.name`

E.g. I want to get the list of `/updates` objects composed of `status` and `updateId`.

`&fields=updateId,status`

E.g. I want all public fields of an update except `enqueudAt`

`&fields=-enqueudAt`

By default `fields` is equivalent to `*` in order to display all public fields of an object.

**Comment by gmourier** [comment]
# Response body format

## Null and Empty
TBD

## Date and Time

🟢 We specify the date and time with the `ISO 8601` standard. Example of what we use as a value in an `update` object for the `processedAt` field. `2021-07-20T20:17:58.432491Z`

## Body Response blocks

#### Multiple objects response

🟢 All multiple objects responses must be represented by a JSON object. That is, responses containing a list of items are now placed in an array called `data` belonging to the root response object.

This allows us to add additional response parameters for paging, hateoas, virtually computed field, etc.. outside of the results dataset.

```
{
  "duration": //duration sum of attribute in data items. This is just an example.
  "data": [
    {
      "status": "processed",
      "updateId": 0,
      "type": {
          "name": "DocumentsPartial",
          "number": 19546
      },
      "duration": 17.405,
      "enqueuedAt": "2021-07-20T20:17:41.026784Z",
      "processedAt": "2021-07-20T20:17:58.432491Z"
    },
    ...
  ],
  "count": 10, //example
  "limit": 10, //example
  "offset": 0 //example
}
```

#### Single object response

🟢 All single object responses must be represented by a JSON object.

```
{
  "status": "processed",
  "updateId": 0,
  "type": {
      "name": "DocumentsPartial",
      "number": 19546
  },
  "duration": 17.405,
  "enqueuedAt": "2021-07-20T20:17:41.026784Z",
  "processedAt": "2021-07-20T20:17:58.432491Z"
}
```

**Comment by gmourier** [comment]
# Custom Headers

🟢 MeiliSearch use custom headers prefixed with `X-Meili-` to follow the [RFC 2047](https://www.ietf.org/rfc/rfc2047.txt).

E.g. `X-Meili-API-Key`

**Comment by gmourier** [comment]
# Query parameter format for array

🟢 Values are separated by `,` e.g. `&attributesToRetrieve=title,author`

> This means that we won't handle handling defined user attributes containing a `,` char in URL query parameters.
> This means that we won't handle any sort of format different from the previous example.

# Request body parameter format for array

🟢 Won't handle array as a string separated by a reserved char when we can use a real JSON type.

> It means that `"attributesToRetrieve": "title,author"` is considered invalid.

**Comment by gmourier** [comment]
# 🟢 A convention for the `status` field, shared on all our APIs.

It is important to define at least an enumeration basis of the different statuses used by the different objects of the same API having an evolving state through time. It is even better when it is shared by all our APIs, thus reinforcing the ease of understanding of our different APIs because they share a common language.

## Update `status` vs Dump `status`

### Update case

| old        | new        |
|------------|------------|
| enqueued   | -   |
| processing | - |
| processed  | **succeeded**  |
| failed     | -     |

### Dump case

| old         | new        |
|-------------|------------|
| in_progress | **processing** |
| done        | **succeeded**  |
| failed      |  -         |

The status of the different objects shares the names that mean similar states. The understanding is improved because they share a common language as much as possible.

`status`  field for `dump` and `update` objects is now sharing similar definitions like `processing` and `succeeded`.

**Comment by gmourier** [comment]
Closed with the release of version [v0.28.0](https://github.com/meilisearch/meilisearch/releases/tag/v0.28.0)
